containerd-shim-spin/src/engine.rs

```diff
@@ -513,6 +517,21 @@ impl ResolvedAppSource {
         ensure!(!types.is_empty(), "no triggers in app");
         Ok(types.into_iter().map(|t| t.as_str()).collect())
     }
+
+    pub fn variables(&self) -> Vec<&str> {
+        match self {
+            ResolvedAppSource::File { manifest, .. } => manifest
+                .variables
+                .keys()
+                .map(|k| k.as_ref())
+                .collect::<Vec<_>>(),
+            ResolvedAppSource::OciRegistry { locked_app } => locked_app
+                .variables
+                .keys()
+                .map(|k| k.as_ref())
+                .collect::<Vec<_>>(),
+        }
+    }
```

Comment:
**suggestion**: refactoring

```suggestion
    pub fn variables(&self) -> Vec<&str> {
        match self {
            ResolvedAppSource::File { manifest, .. } => manifest.variables.keys(),
            ResolvedAppSource::OciRegistry { locked_app } => locked_app.variables.keys(),
        }
        .map(AsRef::as_ref)
        .collect()
    }
```

Reply:
I was hoping to avoid collecting the iterator here all together however the `Keys` are different for the two iterators so you cannot combine them into one iterator:
```sh
note: `Keys<'_, String, Variable>` and `Keys<'_, Id<'_', true>, ...>` have similar names, but are actually distinct types
note: `Keys<'_, String, Variable>` is defined in crate `alloc`
```

Reply:
Ah I see, thanks!